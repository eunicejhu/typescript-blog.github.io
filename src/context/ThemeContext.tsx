import React from "react";

export const LIGHT = {
    text: "rgba(0, 0, 0, 1)",
    background: "rgba(255, 255, 255, 1)",
    btnBackground: "transparent",
    btnBorder: "rgba(0, 0, 0, 1)",
    btnTextHover: "rgba(255, 255, 255, 0.8)",
    btnBackgroundHover: "rgba(0, 0, 0, 0.8)",
    disabled: "rgba(0, 0, 0, 0.24)",
    disabledBackground: "rgba(255, 255, 255, 0.2)",
};
export const DARK = {
    text: "rgba(255, 255, 255, 1)",
    background: "rgba(0, 0, 0, 1)",
    btnBackground: "transparent",
    btnBorder: "rgba(255, 255, 255, 1)",
    btnTextHover: "rgba(0, 0, 0, 0.8)",
    btnBackgroundHover: "rgba(255, 255, 255, 0.8)",
    disabled: "rgba(255, 255, 255, 0.4)",
    disabledBackground: "rgba(0, 0, 0, 0.2)",
};

const PINK = {
    text: "rgba(196, 139, 159, 1)",
    background: "rgba(255, 255, 255, 1)",
    btnBackground: "transparent",
    btnBorder: "rgba(196, 139, 159, 1)",
    btnTextHover: "rgba(255, 255, 255, 0.8)",
    btnBackgroundHover: "rgba(196, 139, 159, 0.8)",
    disabled: "rgba(196, 139, 159, 0.4)",
    disabledBackground: "rgba(196, 139, 159, 0.2)",
};

const OTHER = {
    borderRadius: "4px",
    transition: "all 0.2s",
    spacing: "10px",
    width: "100%",
};

export const themes = {
    dark: { ...DARK, ...OTHER },
    light: { ...LIGHT, ...OTHER },
    pink: { ...PINK, ...OTHER },
};

export type Mode = "pink" | "dark" | "light";
export type Themes = typeof themes;
export type SetMode = (mode: Mode) => void;

const ThemeContext = React.createContext<{
    mode: Mode;
    themes: Themes;
    setMode: SetMode;
}>({
    mode: "pink",
    themes: themes,
    setMode: () => {},
});

ThemeContext.displayName = "ThemeContext";

export default ThemeContext;
