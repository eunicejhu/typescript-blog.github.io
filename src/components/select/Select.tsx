import React, { useState, useEffect, useCallback } from "react";
import styled from "styled-components";
import { Themes, Mode } from "../../context/ThemeContext";
export interface SelectProps {
    themes: Themes;
    mode: Mode;
    data: { id: string; name: string }[];
    placeholder: string;
    width?: string;
    [p: string]: any;
}
const StyledSelect = styled.div<SelectProps>`
    ${({ themes, mode, ...restProps }) => {
        return `
            &&& {
                width: ${restProps.width || themes[mode].width};
                position: relative;
            }
            &&& input {
                padding: ${themes[mode].spacing};
                border: 1px solid ${themes[mode].btnBorder};
                border-radius: ${themes[mode].borderRadius};
                background-color: ${themes[mode].background};
                width: 100%;
                box-sizing: border-box;
                outline: none;
                color: transparent;
                text-shadow: 0 0 0 ${themes[mode].text};

                &:hover {
                    cursor: pointer;
                }
            }
            &&& ul {
                &.open {
                    display: block
                    z-index: 2;
                    background-color: ${themes[mode].background};
                }
                &.close {
                    display: none;
                }
                position: absolute;
                background-color: ${themes[mode].btnBackground};
                border:1px solid ${themes[mode].btnBorder};
                width: 100%;
                box-sizing: border-box;
                margin: 0;
                padding: 0;

                li {
                    list-style: none;
                    color: ${themes[mode].btnBorder};
                    padding: ${themes[mode].spacing};
                    &:hover {
                        background-color: ${themes[mode].btnBackgroundHover};
                        color: ${themes[mode].btnTextHover};
                        cursor: pointer;
                    }
                }
            }
        `;
    }}
`;

const useClickOutside = (cb: () => void) => {
    const outsideClickHandler = useCallback(
        (e: Event) => {
            if (
                ["selectInput", "selectInput-li"].indexOf(
                    (e.target as HTMLElement).classList[0]
                ) === -1
            )
                cb();
        },
        [cb]
    );

    useEffect(() => {
        document.addEventListener("click", outsideClickHandler);
        return () => {
            document.removeEventListener("click", outsideClickHandler);
        };
    }, [outsideClickHandler]);
};

const Select: React.FC<SelectProps> = ({
    data,
    placeholder,
    onChange,
    value,
    ...props
}) => {
    const [focus, setFocus] = useState<boolean>(false);

    useClickOutside(setFocus.bind(this, false));

    const onSelectFocused = () => {
        setFocus(true);
    };
    const onLiClicked = (e: React.FormEvent) => {
        const id = (e.target as HTMLLIElement).dataset.id;
        onChange(id as string);
        setFocus(false);
    };

    const datalist = data.map((item) => {
        return (
            <li
                data-testid={item.id}
                data-id={item.id}
                data-name={item.name}
                className="selectInput-li"
                key={item.id}
                onClick={onLiClicked}
            >
                {item.name}
            </li>
        );
    });
    return (
        /** Here we use uncontrolled component, because it's onClick of li
            charging the change of the value, instead of onChange of input. */
        <StyledSelect {...props}>
            <input
                data-testid="selectInput"
                className="selectInput"
                onFocus={onSelectFocused}
                type="text"
                defaultValue={value}
                placeholder={placeholder}
            ></input>
            <ul className={focus ? "open" : "close"}>{datalist}</ul>
        </StyledSelect>
    );
};

export default Select;
