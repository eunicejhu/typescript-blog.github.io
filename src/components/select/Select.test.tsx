import React from "react";
import { render } from "@testing-library/react";
import Select from "./Select";
import { themes } from "../../context/ThemeContext";
import { fireEvent } from "@testing-library/react";

export const data = [
    {
        id: "1",
        name: "Chris",
    },
    {
        id: "3",
        name: "Isa",
    },
    {
        id: "2",
        name: "Bob",
    },
];

describe("Select test", () => {
    it("Select test:", () => {
        const ui = (
            <Select
                themes={themes}
                mode="pink"
                data={data}
                placeholder="choose the user"
            ></Select>
        );
        const { container } = render(ui);
        expect(container.firstChild).toMatchSnapshot();
    });

    it("select li, update input value", () => {
        const mockOnChange = jest.fn();
        const ui = (
            <Select
                themes={themes}
                mode="pink"
                data={data}
                placeholder="choose the user"
                value=""
                onChange={mockOnChange}
            ></Select>
        );
        const { getByTestId } = render(ui);
        fireEvent.click(getByTestId("1"));
        expect(mockOnChange).toHaveBeenCalledWith("1");
    });
});
