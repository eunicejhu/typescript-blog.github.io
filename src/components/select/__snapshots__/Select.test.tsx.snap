// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select test Select test: 1`] = `
<div
  class="sc-bdfBwQ GIJTt"
  mode="pink"
>
  <input
    class="selectInput"
    data-testid="selectInput"
    placeholder="choose the user"
    type="text"
    value=""
  />
  <ul
    class="close"
  >
    <li
      class="selectInput-li"
      data-id="1"
      data-name="Chris"
      data-testid="1"
    >
      Chris
    </li>
    <li
      class="selectInput-li"
      data-id="3"
      data-name="Isa"
      data-testid="3"
    >
      Isa
    </li>
    <li
      class="selectInput-li"
      data-id="2"
      data-name="Bob"
      data-testid="2"
    >
      Bob
    </li>
  </ul>
</div>
`;
